### ISSUE TITLE
Hello, I trained on batchsize 2048 and got epoch_13_iter_3999.bin.

### ISSUE THREAD

**Issue by leo23ui** [microsoft/Cream#252]
Hello, I got epoch_13_iter_3999.bin when batchsize was 2048. Then I changed batchsize to 1024 and continued training in epoch_13_iter_3999.bin. The batch_count value has exceeded num_batches_per_epoch_r: Train Epoch: 13 [16011/14656] [16395264/15007744 (109%)], but do not enter epoch14 training. Hope to get your advice, thanks!!!
      
could i just change the following code in train.py, thanks!!!

       #if num_feed_images >= all_num_feed_images: 
            #break
        if batch_count >= num_batches_per_epoch:
            break

**Comment by mnmnk43434** [comment]
Hello, we tried to solve the issue.

This is what we did:

Modified the train_epoch function to reset the batch count and move to the next epoch when the number of processed batches exceeds the total number of batches per epoch. This ensures that the training continues to the next epoch correctly, even when changing batch sizes mid-training.

You can review changes in this commit: https://github.com/mnmnk43434/microsoft-Cream-252/commit/b09910cdba09d27c0751aa5750c869d3af18249e.


> [!CAUTION]
> Disclaimer: The concept of solution was created by AI and you should never copy paste this code before you check the correctness of generated code. Solution might not be complete, you should use this code as an inspiration only.

---

If you no longer want Latta AI to attempt solving issues on your repository, you can block this account.

**Comment by leo23ui** [comment]
> Hello, we tried to solve the issue.


thanks very much!!!
 i just change the following code in train.py
  #if num_feed_images >= all_num_feed_images: 
        #break
    if batch_count >= num_batches_per_epoch:
        break


But it report the following warning that didnot happen before


2024-12-08,22:10:14 | INFO | Train Epoch: 14 [41/14656] [   41984/15007744 (0%)] Loss: 3.6008 (3.656) gpu_memory: 7.0049 (7.0049)       cpu_memory: 56.8142 (56.8142)   rest_shm: 747.5300 (747.5300)   data_time: 0.1891 (0.4806)     soft_img2text_loss: 1.8171 (1.8381)     soft_text2img_loss: 1.7837 (1.8197)     soft_loss: 3.6008 (3.6578)      loss_scale: 131072.0000 (131072.0000)   batch_time: 6.8644 (7.1550)     throughput: 149.1750 (146.7896)         LR: 0.000041 Logit Scale: 50.000
wandb: WARNING (User provided step: 205224 is less than current step: 205528. Dropping entry: {'train/loss': 3.600809097290039, 'step': 205224, 'num_feed_images': 210150400, '_timestamp': 1733667014.6765273}).
wandb: WARNING (User provided step: 205224 is less than current step: 205528. Dropping entry: {'train/scale': 50.000003814697266, 'step': 205224, 'num_feed_images': 210150400, '_timestamp': 1733667014.6766698}).
wandb: WARNING (User provided step: 205224 is less than current step: 205528. Dropping entry: {'train/lr': 4.098685672230562e-05, 'step': 205224, 'num_feed_images': 210150400, '_timestamp': 1733667014.6767316}).
wandb: WARNING (User provided step: 205224 is less than current step: 205528. Dropping entry: {'train/lr_l0': 4.098685672230562e-05, 'step': 205224, 'num_feed_images': 210150400, '_timestamp': 1733667014.6767929}).
wandb: WARNING (User provided step: 205224 is less than current step: 205528. Dropping entry: {'train/gpu_memory': 7.0048828125, 'step': 205224, 'num_feed_images': 210150400, '_timestamp': 1733667014.6768482}).
wandb: WARNING (User provided step: 205224 is less than current step: 205528. Dropping entry: {'train/cpu_memory': 56.814205169677734, 'step': 205224, 'num_feed_images': 210150400, '_timestamp': 1733667014.6769018}).
wandb: WARNING (User provided step: 205224 is less than current step: 205528. Dropping entry: {'train/rest_shm': 747.5300483703613, 'step': 205224, 'num_feed_images': 210150400, '_timestamp': 1733667014.6769528}).
wandb: WARNING (User provided step: 205224 is less than current step: 205528. Dropping entry: {'train/data_time': 0.18909382820129395, 'step': 205224, 'num_feed_images': 210150400, '_timestamp': 1733667014.6770039}).
wandb: WARNING (User provided step: 205224 is less than current step: 205528. Dropping entry: {'train/soft_img2text_loss': 1.8170948028564453, 'step': 205224, 'num_feed_images': 210150400, '_timestamp': 1733667014.677054}).
wandb: WARNING (User provided step: 205224 is less than current step: 205528. Dropping entry: {'train/soft_text2img_loss': 1.7837142944335938, 'step': 205224, 'num_feed_images': 210150400, '_timestamp': 1733667014.6771061}).
wandb: WARNING (User provided step: 205224 is less than current step: 205528. Dropping entry: {'train/soft_loss': 3.600809097290039, 'step': 205224, 'num_feed_images': 210150400, '_timestamp': 1733667014.6771564}).
wandb: WARNING (User provided step: 205224 is less than current step: 205528. Dropping entry: {'train/loss_scale': 131072.0, 'step': 205224, 'num_feed_images': 210150400, '_timestamp': 1733667014.677206}).
wandb: WARNING (User provided step: 205224 is less than current step: 205528. Dropping entry: {'train/batch_time': 6.864420652389526, 'step': 205224, 'num_feed_images': 210150400, '_timestamp': 1733667014.6772544}).
wandb: WARNING (User provided step: 205224 is less than current step: 205528. Dropping entry: {'train/throughput': 149.1750071644491, 'step': 205224, 'num_feed_images': 210150400, '_timestamp': 1733667014.677303}).
2024-12-08,22:11:23 | INFO | Train Epoch: 14 [51/14656] [   52224/15007744 (0%)] Loss: 3.7561 (3.673) gpu_memory: 7.0049 (7.0049)       cpu_memory: 56.8142 (56.8142)   rest_shm: 747.5300 (747.5300)   data_time: 0.1850 (0.4231)     soft_img2text_loss: 1.8866 (1.8367)     soft_text2img_loss: 1.8694 (1.8192)     soft_loss: 3.7561 (3.6558)      loss_scale: 131072.0000 (131072.0000)   batch_time: 6.8602 (7.0977)     throughput: 149.2661 (147.2651)         LR: 0.000041 Logit Scale: 50.000
wandb: WARNING (User provided step: 205234 is less than current step: 205528. Dropping entry: {'train/loss': 3.756077527999878, 'step': 205234, 'num_feed_images': 210160640, '_timestamp': 1733667083.3025541}).
wandb: WARNING (User provided step: 205234 is less than current step: 205528. Dropping entry: {'train/scale': 50.000003814697266, 'step': 205234, 'num_feed_images': 210160640, '_timestamp': 1733667083.3027163}).
wandb: WARNING (User provided step: 205234 is less than current step: 205528. Dropping entry: {'train/lr': 4.09826167330748e-05, 'step': 205234, 'num_feed_images': 210160640, '_timestamp': 1733667083.3027844}).
wandb: WARNING (User provided step: 205234 is less than current step: 205528. Dropping entry: {'train/lr_l0': 4.09826167330748e-05, 'step': 205234, 'num_feed_images': 210160640, '_timestamp': 1733667083.3028517}).
wandb: WARNING (User provided step: 205234 is less than current step: 205528. Dropping entry: {'train/gpu_memory': 7.0048828125, 'step': 205234, 'num_feed_images': 210160640, '_timestamp': 1733667083.3029125}).
wandb: WARNING (User provided step: 205234 is less than current step: 205528. Dropping entry: {'train/cpu_memory': 56.814205169677734, 'step': 205234, 'num_feed_images': 210160640, '_timestamp': 1733667083.3029704}).
wandb: WARNING (User provided step: 205234 is less than current step: 205528. Dropping entry: {'train/rest_shm': 747.5300483703613, 'step': 205234, 'num_feed_images': 210160640, '_timestamp': 1733667083.3030267}).
wandb: WARNING (User provided step: 205234 is less than current step: 205528. Dropping entry: {'train/data_time': 0.18501019477844238, 'step': 205234, 'num_feed_images': 210160640, '_timestamp': 1733667083.303078}).
wandb: WARNING (User provided step: 205234 is less than current step: 205528. Dropping entry: {'train/soft_img2text_loss': 1.8866482973098755, 'step': 205234, 'num_feed_images': 210160640, '_timestamp': 1733667083.303128}).
wandb: WARNING (User provided step: 205234 is less than current step: 205528. Dropping entry: {'train/soft_text2img_loss': 1.8694292306900024, 'step': 205234, 'num_feed_images': 210160640, '_timestamp': 1733667083.303179}).
wandb: WARNING (User provided step: 205234 is less than current step: 205528. Dropping entry: {'train/soft_loss': 3.756077527999878, 'step': 205234, 'num_feed_images': 210160640, '_timestamp': 1733667083.3032277}).
wandb: WARNING (User provided step: 205234 is less than current step: 205528. Dropping entry: {'train/loss_scale': 131072.0, 'step': 205234, 'num_feed_images': 210160640, '_timestamp': 1733667083.3032763}).
wandb: WARNING (User provided step: 205234 is less than current step: 205528. Dropping entry: {'train/batch_time': 6.860230922698975, 'step': 205234, 'num_feed_images': 210160640, '_timestamp': 1733667083.3033237}).
wandb: WARNING (User provided step: 205234 is less than current step: 205528. Dropping entry: {'train/throughput': 149.26611240035845, 'step': 205234, 'num_feed_images': 210160640, '_timestamp': 1733667083.303373}).

**Comment by wkcn** [comment]
Hi @leo23ui , when the batch size is changed from 2048 to 1024, the iteration should be changed (star_iter *= 2).

start_iter:  https://github.com/microsoft/Cream/blob/main/TinyCLIP/src/training/main.py#L401
```
start_iter = checkpoint['iter_in_epoch'] * 2 + 1
```

**Comment by leo23ui** [comment]
> Hi [@leo23ui](https://github.com/leo23ui) , when the batch size is changed from 2048 to 1024, the iteration should be changed (star_iter *= 2).
> 
> start_iter: https://github.com/microsoft/Cream/blob/main/TinyCLIP/src/training/main.py#L401
> 
> ```
> start_iter = checkpoint['iter_in_epoch'] * 2 + 1
> ```

thanks for your reply very much!!!

**Comment by leo23ui** [comment]
> Hi [@leo23ui](https://github.com/leo23ui) , when the batch size is changed from 2048 to 1024, the iteration should be changed (star_iter *= 2).


Hello, I got epoch_13_iter_3999.bin when batchsize was 2048. Then I changed batchsize to 1024 and continued training in epoch_13_iter_3999.bin.       

And i changed the start_iter = checkpoint['iter_in_epoch']* 2 + 1

But after i run the train sh based on epoch_13_iter_3999.bin,  it start the evaluation on imagenet, but after the evaluation, it save the model epoch_13_iter_7999.bin rather than epoch_13_iter_4999.bin. Will this affect the training of the model? Looking forward to your reply, thanks!!!
